### ISSUE TITLE
[complex_autograd] [TorchScript] Autograd graph is lost on 2nd run when using complex tensors with @torch.jit.script

### ISSUE THREAD

**Issue by BouazzaSE** [pytorch/pytorch#65480]
## 🐛 Bug

When running **for a second time** a TorchScript function which works on complex tensors (among them is an input float Tensor which we turn into a complex tensor with .cfloat inside the function), the autograd graph is lost, ie the result has no `grad_fn`.  It looks like somehow torch.jit's graph optimizer thinks it's a good idea to disable gradients after the second pass. Removing the `@torch.jit.script` decorator, everything works as expected.

## To Reproduce

1. Run the code below:
```python
import torch

def foo(x: torch.Tensor, y: torch.Tensor, W: torch.Tensor):
    return torch.exp(torch.mm(torch.complex(x, y), W.cfloat()))

@torch.jit.script
def jitted_foo(x: torch.Tensor, y: torch.Tensor, W: torch.Tensor):
    return torch.exp(torch.mm(torch.complex(x, y), W.cfloat()))

x = torch.randn(128, 16, dtype=torch.float32, device='cuda:0')
y = torch.randn(128, 16, dtype=torch.float32, device='cuda:0')
W = torch.randn(16, 1, dtype=torch.float32, device='cuda:0', requires_grad=True)
W.data /= 4
print('1st launch - foo(...).grad_fn is None = {}'.format(foo(x, y, W).grad_fn is None)) # outputs False, which is what's expected
print('2nd launch - foo(...).grad_fn is None = {}'.format(foo(x, y, W).grad_fn is None)) # outputs False, again as expected
print('1st launch - jitted_foo(...).grad_fn is None = {}'.format(jitted_foo(x, y, W).grad_fn is None)) # outputs False, again as expected
print('2nd launch - jitted_foo(...).grad_fn is None = {}'.format(jitted_foo(x, y, W).grad_fn is None)) # outputs True, meaning the graph is lost
```

## Expected behavior
All of the prints should show "False", ie `jitted_foo(x, y, W).grad_fn` should NOT be None.

## Environment
Output of the [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py):
```
PyTorch version: 1.9.0a0+gitd69c22d
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.19.6
Libc version: glibc-2.28

Python version: 3.8.11 (default, Aug  3 2021, 15:09:35)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.19.0-9-amd64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 450.51.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.10.0a0+gitf69cf3c
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.0.221             h6bb024c_0
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2021.3.0           h06a4308_520
[conda] mkl-include               2021.3.0           h06a4308_520
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.0            py38h42c9631_2
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.20.3           py38hf144106_0
[conda] numpy-base                1.20.3           py38h74d4b33_0
[conda] torch                     1.9.0a0+gitd69c22d          pypi_0    pypi
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved

**Comment by BouazzaSE** [comment]
By the way, tracing works:
```python
traced_foo = torch.jit.trace(foo, (x, y, W.detach()))
print('1st launch - traced_foo(...).grad_fn is None = {}'.format(traced_foo(x, y, W).grad_fn is None)) # outputs False, as expected
print('2nd launch - traced_foo(...).grad_fn is None = {}'.format(traced_foo(x, y, W).grad_fn is None)) # outputs False, as expected
```

**Comment by BouazzaSE** [comment]
replaced 
```python
W = torch.randn(16, 1, dtype=torch.float32, device='cuda:0', requires_grad=True)
```
with
```python
W = torch.randn(16, 1, dtype=torch.float32, device='cuda:0', requires_grad=True)
W.data /= 4
```
for more correctness as the previous version would have caused errors with `.backward`, though we aren't using backward and it still doesn't change the fact that `grad_fn` still turns to None on the second launch of the scripted function.

**Comment by BouazzaSE** [comment]
I found a fix that works and makes the jitted function behave exactly like the traced one now. Simply do a pass in no_grad mode before attempting to use the jitted function:
```python
import torch

def foo(x: torch.Tensor, y: torch.Tensor, W: torch.Tensor):
    return torch.exp(torch.mm(torch.complex(x, y), W.cfloat()))

@torch.jit.script
def jitted_foo(x: torch.Tensor, y: torch.Tensor, W: torch.Tensor):
    return torch.exp(torch.mm(torch.complex(x, y), W.cfloat()))

x = torch.randn(128, 16, dtype=torch.float32, device='cuda:0')
y = torch.randn(128, 16, dtype=torch.float32, device='cuda:0')
W = torch.randn(16, 1, dtype=torch.float32, device='cuda:0', requires_grad=True)
W.data /= 4
# BEGIN WORKAROUND FOR jitted_foo
with torch.no_grad():
    jitted_foo(x, y, W)
# END WORKAROUND FOR jitted_foo
print('1st launch - foo(...).grad_fn is None = {}'.format(foo(x, y, W).grad_fn is None)) # outputs False, which is what's expected
print('2nd launch - foo(...).grad_fn is None = {}'.format(foo(x, y, W).grad_fn is None)) # outputs False, again as expected
print('1st launch - jitted_foo(...).grad_fn is None = {}'.format(jitted_foo(x, y, W).grad_fn is None)) # outputs False, again as expected
print('2nd launch - jitted_foo(...).grad_fn is None = {}'.format(jitted_foo(x, y, W).grad_fn is None)) # outputs True, meaning the graph is lost
traced_foo = torch.jit.trace(foo, (x, y, W.detach()))
print('1st launch - traced_foo(...).grad_fn is None = {}'.format(traced_foo(x, y, W).grad_fn is None)) # outputs False, as expected
print('2nd launch - traced_foo(...).grad_fn is None = {}'.format(traced_foo(x, y, W).grad_fn is None)) # outputs False, as expected
```
Now all of the grad_fn's are not None, which is what we expect.
The graph for `jitted_foo` also changed now from:
```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %W.1 : Tensor):
  %5 : int = prim::Constant[value=9]() # /tmp/ipykernel_31892/307437112.py:8:51
  %4 : bool = prim::Constant[value=0]()
  %3 : NoneType = prim::Constant()
  %6 : ComplexFloat(128, 16, strides=[16, 1], requires_grad=0, device=cuda:0) = aten::complex(%x.1, %y.1) # /tmp/ipykernel_31892/307437112.py:8:30
  %8 : ComplexFloat(16, 1, strides=[1, 1], requires_grad=1, device=cuda:0) = aten::to(%W.1, %5, %4, %4, %3) # /tmp/ipykernel_31892/307437112.py:8:51
  %30 : ComplexFloat(16, 1, strides=[1, 1], requires_grad=1, device=cuda:0), %31 : ComplexFloat(128, 16, strides=[16, 1], requires_grad=0, device=cuda:0), %32 : bool = prim::RequiresGradCheck[types=[Tensor(requires_grad=1), Tensor(requires_grad=0)]](%8, %6)
  %33 : Tensor = prim::If(%32)
    block0():
      %16 : Tensor = prim::DifferentiableGraph_0(%30, %31)
      -> (%16)
    block1():
      %41 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %42 : (Tensor) = prim::CallFunction(%41, %8, %6)
      %43 : Tensor = prim::TupleUnpack(%42)
      -> (%43)
  return (%33)
with prim::DifferentiableGraph_0 = graph(%9 : Tensor(requires_grad=0),
      %11 : Tensor(requires_grad=0)):
  %8 : Tensor(requires_grad=0) = aten::mm(%11, %9) # /tmp/ipykernel_31892/307437112.py:8:21
  %3 : Tensor = aten::exp(%8) # /tmp/ipykernel_31892/307437112.py:8:11
  return (%3, %11, %9, %3)
```
to
```
graph(%x.1 : Tensor,
      %y.1 : Tensor,
      %W.1 : Tensor):
  %3 : int = prim::Constant[value=9]() # /tmp/ipykernel_31892/1212327360.py:8:51
  %4 : bool = prim::Constant[value=0]()
  %5 : NoneType = prim::Constant()
  %6 : Tensor = aten::complex(%x.1, %y.1) # /tmp/ipykernel_31892/1212327360.py:8:30
  %7 : Tensor = aten::to(%W.1, %3, %4, %4, %5) # /tmp/ipykernel_31892/1212327360.py:8:51
  %8 : Tensor = aten::mm(%6, %7) # /tmp/ipykernel_31892/1212327360.py:8:21
  %9 : Tensor = aten::exp(%8) # /tmp/ipykernel_31892/1212327360.py:8:11
  return (%9)
```
which exactly matches the graph of `traced_foo`.
